httpx/cert_reload.go

```diff
@@ -0,0 +1,159 @@
+)
+
+type (
+	CertLocation struct {
+		CertPath string
+		KeyPath  string
+	}
+
+	CertLoadFunc func() ([]tls.Certificate, error)
+
+	CertProvider struct {
+		certs   []tls.Certificate
+		mu      sync.Mutex
+		watcher *fsnotify.Watcher
+
+		certLoad     CertLoadFunc
+		certLocation CertLocation
+		logger       *logrusx.Logger
+	}
+)
+
+//NewCertProvider creates a tls.Certificate provider that reloads certificates on fs change
+//certs must not be empty
+func NewCertProvider(certs []tls.Certificate, certLocation CertLocation, certLoad CertLoadFunc, l *logrusx.Logger) *CertProvider {
+	ret := &CertProvider{
+		certLoad:     certLoad,
+		certLocation: certLocation,
+		logger:       l,
+	}
+	ret.load(certs)
+	ret.watchCertificatesChanges()
+
+	runtime.SetFinalizer(ret, func(ret *CertProvider) { ret.Stop() })
+
+	return ret
+}
+
+func (p *CertProvider) load(certs []tls.Certificate) {
+	for i := range certs {
+		tlsCert := &certs[i]
+		if tlsCert.Leaf != nil {
+			continue
+		}
+		for _, bCert := range tlsCert.Certificate {
+			cert, err := x509.ParseCertificate(bCert)
+			if err == nil && !cert.IsCA {
+				tlsCert.Leaf = cert
+			}
+		}
+	}
+
+	p.mu.Lock()
+	defer p.mu.Unlock()
+	p.certs = certs
+}
+
+func (p *CertProvider) watchCertificatesChanges() {
+	var err error
+	p.watcher, err = fsnotify.NewWatcher()
+	if err != nil {
+		p.logger.WithError(err).Fatalf("Could not activate certificate change watcher")
+	}
+
+	go func() {
+		p.logger.Infof("Starting tls certificate auto-refresh")
+		for {
+			select {
+			case _, ok := <-p.watcher.Events:
+				if !ok {
+					return
+				}
+
+				p.waitForAllFilesChanges()
+
+				p.logger.Infof("TLS certificates changed, updating")
+				certs, err := p.certLoad()
+				if err != nil {
+					p.logger.WithError(err).Errorf("Error in the new tls certificates")
+					return
+				}
+				p.load(certs)
+			case err, ok := <-p.watcher.Errors:
+				if !ok {
+					return
+				}
+				p.logger.WithError(err).Errorf("Error occured in the tls certificate change watcher")
+			}
+		}
+	}()
+
+	certPath := path.Dir(p.certLocation.CertPath)
+	keyPath := path.Dir(p.certLocation.KeyPath)
+
+	err = p.watcher.Add(certPath)
+	if err != nil {
+		p.logger.WithError(err).Errorf("Error watching the certFolder for tls certificate change")
+	}
+
+	if certPath != keyPath {
+		err = p.watcher.Add(keyPath)
+		if err != nil {
+			p.logger.WithError(err).Errorf("Error watching the keyFolder for tls certificate change")
+		}
+	}
+}
+
+func (p *CertProvider) waitForAllFilesChanges() {
+	flushUntil := time.After(2 * time.Second)
+	p.logger.Infof("TLS certificates files changed, waiting for changes to finish")
+	stop := false
+	for {
+		select {
+		case <-flushUntil:
+			stop = true
+		case <-p.watcher.Events:
+			continue
+		case <-p.watcher.Errors:
+			continue
+		}
+
+		if stop {
+			break
+		}
+	}
+}
+
+func (p *CertProvider) GetCertificate(hello *tls.ClientHelloInfo) (*tls.Certificate, error) {
+	p.mu.Lock()
+	defer p.mu.Unlock()
```

Comment:
If you want to keep this mutex as opposed to switching to sync/atomic, I would propose this to reduce contention:

```suggestion
	certs := p.certs
	p.mu.Unlock()
```

and use `certs` instead of `p.certs` in the following.